### ISSUE TITLE
Make the armor coverage roll overlap, instead of pure chance

### ISSUE THREAD

**Issue by col-a-guo** [cataclysmbnteam/Cataclysm-BN#1608]
### Is your feature request related to a problem? Please describe.

Right now if you're wearing, say, platemail and chainmail, you have 95% coverage on both the normal and outer layers. However, a "bullet skip" sometimes occurs where you roll that 5% chance to ignore on both plate AND chain; this can result in you getting oneshot while, say, throwing EMP grenades at Talon UGVs in the lab, when you should have survived 99.75% of the time.

### Describe the solution you'd like

Carry the RNG through the coverage calculation process. If I understand correctly, this means you'd add the RNG variable in absorb_hit before the body part loop, and carry it into armor_absorb's arguments. Then armor_absorb would mutate both "RNG" and "total coverage" to track hits.

So if you have 90%, 70%, 10%, 50% on the same body part, and you roll a 50, it goes:

RNG = 50 < 90 = total coverage, block with item 1
(add 100 to RNG, add next item's coverage)
RNG = 150 < 160 = total coverage, block with item 2
(add 100 to RNG, add next item's coverage)
RNG = 250 > 170 = total coverage, don't block with item 3
(add next item's coverage)
RNG = 250 > 220 = total coverage, don't block with item 4

If you rolled a 10 instead, the last comparison would be 210 < 220 and you'd block with item 4, too.

If you rolled a 65 instead, you wouldn't block with item 2 (165 > 160), but you'd end up blocking with item 3 instead (165 < 220).

If you rolled a 75, similarly, you wouldn't block with item 2 (175 > 160), or item 3 (175 > 170), but you'd block with item 4 (175 < 220

### Describe alternatives you've considered

First, alternatives mentioned include making damage reduction % a universal nonstacking modifier on some armor, and/or creating "tiers" of coverage with distinct features.

Second, to address the main issue brought up with _this_ method, which I proposed as a sort of quick fix in the interim before those other changes (which I should say I agree are better long term). The issue is that this method might either "create a bullet skip" out of two high coverage items - basically that layering armor carefully could be better than doing so haphazardly. However, I've looked at the math, and it's (on a probabilistic level) strictly lower than the current "skip chance" anyway.

First, here's the meta-probability concept. If you imagine that the coverages were totally random, then your average skip chance for two items of a given protection is (1-coverage 1)*(1-coverage 2). Your average skip chance under this new system for adjacent items is 1-coverage 1-coverage 2, which is strictly better for percentages less than 100%.

Now, your average skip chance for non adjacent items, assuming the coverage is totally randomized... is still (1-coverage 1)x(1-coverage 2). Because you're basically trying to line up the two gaps in the coverage - imagine putting a pin on every point from 0 to 100%, and randomly placing a 60% long block and a 70% block along that line. On average, how many pins are you gonna dodge both blocks? (100%-60%)x(100%-70%) = 12%. The same!

However, it's actually way better than that, because you need a specific arrangement (modulo 100%) that turns out to be very rare. Basically you need 2+ intermediate items with coverage totaling right around (200%-first "good" layer's coverage). And given that the layers appear to be:
- Inner, weak, high coverage
- Waist, weak, low coverage
- Normal, strong, high coverage
- Outer, strong, high coverage
- Strapped, varies, variable coverage

The only instance of this I can see being even a little possible is having a weak "outer", and a strong "normal" and "strapped".  Now this isn't impossible:  Notably metal arm and leg guards have 80% coverage, and greaves have 70%. I'm not counting knee/elbow pads because they have 30% and thus are pretty low coverage; you wouldn't call that a "skip". Leather vambraces are 50% which is pretty borderline. You could conceivably have an issue there, except I just checked the entire list of craftable arm and leg gear, and the lowest possible coverage for "outer" is 60% on a Haori exactly (and 65% on cloaks and leather chaps).

So the worst you can possibly do is a Haori for crappy warmth and leather vambraces. That'd put you at 110%; that means if your "normal" armor has 50% coverage, your skip chance is 40%, instead of 25%, if you can call it a skip at that point. And it linearly deceases from there, until 80% "normal" coverage has equal skip chance either way, and then it improves from there.

But leather vambraces are outclassed by hard arm guards for ranged characters (3 less encumbrance, better coverage, 1 more acid resist, in exchange for 1 cut and 1 ballistic defense), and metal arm guards or even scrap arm guards for melee characters (way more defense, better coverage). And the Haori sucks generally. So this seems unrealistic.

What if you jump up to 300%? Then sure, there's a few combos. Worst case, 2 cloaks for warmth leaves you at 130%, and then you have mediocre results with just about any medium-coverage strapped armor; the worst case is greaves + 2 cloaks, which will triple skip chance.

...but why wear two cloaks exactly? If you're going to be super frozen, you're going to want more than that, while if you're just in normal winter weather, a fur duster has better protection and storage, for minimal loss in encumbrance and warmth.

And of course, if you wear anything else, say, a cloak, a duster, and leg guards, then you have zero skip chance.

**TL;DR,** The suggested overlapping process definitely reduces number of skips. It maintains the same number of skips for items with enough space between them to be totally randomized - for other items, because items that can "skip" tend to be high coverage and adjacent, skips go way down.

For the roughest, quickest explanation, imagine if average coverage was 80%. Then on average in order to increase skip rate, you'd need four items between two "good" items. And there just isn't that much stuff, usually.

**Comment by AluminumAlman** [comment]
> So if you have 90%, 70%, 10%, 50% on the same body part, and you roll a 50, it goes:
RNG = 50 < 90 = total coverage, block with item 1
(add 100 to RNG, add next item's coverage)
RNG = 150 < 160 = total coverage, block with item 2
(add 100 to RNG, add next item's coverage)
RNG = 250 > 170 = total coverage, don't block with item 3
(add next item's coverage)
RNG = 250 > 220 = total coverage, don't block with item 4

On case edges, this may cause attacks to always be checked against either the 1st or 2nd piece of clothing if their combined coverage is more than 100% while the 1st piece of clothing has less than 100%. Wouldn't this make it so wearing two ballistic vests cause at least one of them to take the hit to the torso?

**Comment by col-a-guo** [comment]
> > So if you have 90%, 70%, 10%, 50% on the same body part, and you roll a 50, it goes:
> > RNG = 50 < 90 = total coverage, block with item 1
> > (add 100 to RNG, add next item's coverage)
> > RNG = 150 < 160 = total coverage, block with item 2
> > (add 100 to RNG, add next item's coverage)
> > RNG = 250 > 170 = total coverage, don't block with item 3
> > (add next item's coverage)
> > RNG = 250 > 220 = total coverage, don't block with item 4
> 
> On case edges, this may cause attacks to always be checked against either the 1st or 2nd piece of clothing if their combined coverage is more than 100% while the 1st piece of clothing has less than 100%. For example, wearing two ballistic vests would cause at least one of them to take a hit to the torso (between 0 and 80 both vests absorb the hit, between 81 and 90 the 1st one absorbs the hit, between 91 and 100 the 2nd one absorbs the hit). In my opinion, it'd be better to just roll a coverage check for each piece of clothing separately (which is how I'd expect it to be intuitively and how you got the 99.75%).

It's a totally valid opinion to say that "bullet skips" are a _good_ thing; that you _want_ a 0.25% chance to get shot as if you had no armor. That said, the behavior you're describing, where stacking armor provides consistency, is the intended result. That's the goal, basically.

It seems to me, though, that the general consensus is that removing "bullet skips" has more pros than cons; pros include:

- realism (if you're wearing a helmet and goggles, it should always hit one, for example)
- less extreme variance (people want to die because of bad decisions, not bad luck, typically, and from a game design standpoint that's also usually better)
- balance (right now survivor gear is overpowered, in part due to this; in general the game is balanced around a highly swingy exponential curve on the value of coverage). 

The cons that have been discussed with this _particular_ solution include:
- The possibility that this means meta gaming is optimal (see above; it almost always isn't)
- The possibility that you might randomly create an armor set that is _worse_ for bullet skips (see above; it almost always isn't)
- The possibility that relayering gear, a somewhat broken mechanic already, would be optimal (this _usually_ isn't the case, and that's more a problem with relayering, anyway)

The various other suggestions have included: making coverage a specific percentage range (e.g. elbow pads are 40-60) and then roll one single hit, making armor have universal percent damage reduction, etc. 

Notably these other suggestions are probably better long term, at least in theory, in my opinion. However, all the ones I've seen require at _minimum_ changing every item (which is somewhat tricky via automation in the best case, and tremendously time consuming in the worst), if not also creating new mechanics entirely.

This option is useful not because it is comprehensive but because it provides a fast solution that will actually be done immediately; to quote myself:

"I also figured out where to alter it in the code, and it does seem like a ~10 liner; you just roll RNG before the loop (line 8020, character.cpp) instead of in armor_absorb. Then make armor_absorb take two additional arguments ("RNG" and "total_coverage") and mutate the two in armor_absorb as described, i.e. make the existing check:
```
total_coverage += armor.get_coverage()
if( RNG > total_coverage ) {
   return false;
} else {
   RNG += 100

}
```

"

**Comment by col-a-guo** [comment]
Also, I implemented it and it works, I think. https://github.com/col-a-guo/Cataclysm-BN

**Comment by Firestorm01X2** [comment]
@Coolthulhu is that solution of "coverage problem"?

**Comment by Coolthulhu** [comment]
Not really, it's rather complex for the player to do math on it. An unspoiled player could create whole "skip zones" by having two high protection items with a high coverage item in between them due to layering. For example, MBR+t-shirt+kevlar vest.

Some algorithm to best allocate coverage could exist, but it would be more complex than just skipping 100% coverage worth of items between each hit to an item.

While changing it to a coverage% accumulation could be an overall improvement mechanically (I don't know if it is), it's too spoily, and has too much micromanagement for its "fun", to be considered a full solution.
My idea for a full solution would start with treating all items as 100% coverage, then having most of them have a flag like "ignored on armor piercing attack" or "50% armor on armor piercing attack", with crits being treated like armor piercing.

**Comment by col-a-guo** [comment]
@Coolthulhu 

Obviously haven't looked at this for awhile, but:

> While changing it to a coverage% accumulation could be an overall improvement mechanically (I don't know if it is), it's too spoily, and has too much micromanagement for its "fun", to be considered a full solution. My idea for a full solution would start with treating all items as 100% coverage, then having most of them have a flag like "ignored on armor piercing attack" or "50% armor on armor piercing attack", with crits being treated like armor piercing.

I agree this isn't a full solution; this is just a fast solution (and relatively easy; it took me like an hour to figure out the language and 5 minutes to write the dozen or so lines of code). I like fast, minor options - even if not an improvement now, it may become more important later. Still, I understand that's not everyone style, and making too many bandaid solutions can result in bandaid spaghetti (gross).

> An unspoiled player could create whole "skip zones" by having two high protection items with a high coverage item in between them due to layering. For example, MBR+t-shirt+kevlar vest.

I did do the math on this. It's possible but in practice, the chance of large "skip zones" is much less "skipping" overall than the current _certainty_ of small "skip zones".

A more comprehensive solution would probably be to check protection in order of protection amount; then you can't alternate low and high protection to either improve or worsen your odds. That would make spoilers and micromanagement impossible to abuse, and remove "skip zones" entirely. But, that's substantially more work. _If_ that was implemented, would you agree it'd be a full solution?
